pkg/etcdcli/etcdcli.go

```diff
@@ -116,6 +111,9 @@ func (g *etcdClientGetter) getEtcdClient() (*clientv3.Client, error) {
 		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", bootstrapIP))
 	}
 
+	// ensure we always compare in the right order, the endpoints can be returned in any
+	sort.Strings(etcdEndpoints)
+
 	g.clientLock.Lock()
 	defer g.clientLock.Unlock()
```

Comment:
I might be missing something but now that we have `clientLock.RLock()` protecting the client usage in MemberAdd/Remove/List etc... shouldn't we move `g.clientLock.Lock()` down to where we actually modify(close and recreate) the cached client? i.e L131.

Otherwise if goroutine A is in the midst of a `MemberList()` and has acquired `RLock()`, then goroutine B can't get the cached client here when being blocked by `Lock()`.

Reply:
true, this should move down for sure.  
I found sync.Pool https://pkg.go.dev/sync#Pool in the stdlib, which we could use to get a client and return it once done. I reckon that should alleviate all those issues. Not sure how we can close it then - I'll play around with this a bit today.